Generic methods

macsux opened the pull request:
- Generic methods for Kubernetes
- Friend assembly tags
- Timeout handler for when using external HttpClient with Autorest
- Support for bookmarks in the watcher


This PR is pre-req for informers PR https://github.com/kubernetes-client/csharp/pull/394

macsux commented:
/assign @brendandburns

brendandburns commented:
Please resolve conflicts then I will review.

macsux commented:
@brendandburns Resolved

brendandburns commented on src/KubernetesClient/Kubernetes.Generic.cs:
```diff
@@ -0,0 +1,1131 @@
+            var statusCode = httpResponse.StatusCode;
+            cancellationToken.ThrowIfCancellationRequested();
+            string responseContent = null;
+            if ((int)statusCode != 200 && (int)statusCode != 401)
```
This is inconsistent with the `OK` below, please use constants everywhere for consistency.

brendandburns commented on src/KubernetesClient/Kubernetes.Generic.cs:
```diff
@@ -0,0 +1,1131 @@
+            // Set Headers
+
+
+            if (customHeaders != null)
```
A bunch of this logic is repeated in multiple places, please refactor into common methods.

brendandburns commented on src/KubernetesClient/Kubernetes.Generic.cs:
```diff
@@ -0,0 +1,1131 @@
+            // Tracing
+            var shouldTrace = ServiceClientTracing.IsEnabled;
+            string invocationId = null;
+            if (shouldTrace)
```
Though this logic is a little different than previous, I think you could pretty easily refactor it into a helper method that was something like:

```
if (shouldTrace)
{
    addTraces(["exact", exact], ["export", export], ....);
}

brendandburns commented on src/KubernetesClient/Kubernetes.Generic.cs:
```diff
@@ -0,0 +1,1131 @@
+            }
+            var url = GetUrlPathForResource(type, namespaceParameter);
+            var queryParameters = new List<string>();
+            if (allowWatchBookmarks != null)
```
There is logic for building query strings here:

https://github.com/kubernetes-client/csharp/blob/master/src/KubernetesClient/Utilities.cs#L9

That will clean this up a lot.

brendandburns commented on src/KubernetesClient/Kubernetes.Generic.cs:
```diff
@@ -0,0 +1,1131 @@
+
+            if (statusCode == OK)
+            {
+                responseContent = await httpResponse.Content.ReadAsStringAsync().ConfigureAwait(false);
```
I think this logic can be extracted too.

brendandburns commented on src/KubernetesClient/Kubernetes.Generic.cs:
```diff
@@ -0,0 +1,1131 @@
+            var statusCode = httpResponse.StatusCode;
+            cancellationToken.ThrowIfCancellationRequested();
+            string responseContent = null;
+            if (statusCode != OK && statusCode != Accepted && statusCode != Unauthorized)
```
Delete can actually return Accepted and a `Status` object instead of the deleted object. We should probably handle that case here too.

macsux replied:
This code was taken directly from generated methods of Kubernetes object (hence the duplicate code - I wasn't trying to pretty it up beyond making it generic). 

Personally I'm of the opinion that once these methods are standardized, 90% of Kubernetes class autogenerated methods can be made to just call these generic methods.

brendandburns commented on src/KubernetesClient/Extensions.cs:
```diff
@@ -0,0 +1,36 @@
+    public static class Extensions
+    {
+
+        public static KubernetesEntityAttribute GetKubernetesTypeMetadata<T>(this T obj) where T : IKubernetesObject =>
```
We now have an `IMetadata` interface:

https://github.com/kubernetes-client/csharp/blob/master/src/KubernetesClient/IMetadata.cs

I think it makes sense to use that here.

macsux replied:
This is different metadata as in this context I mean grab the type's `KubernetesEntityAttribute` instead of Metadata field. I guess I can rename it to something else - I can see the confusion

brendandburns commented:
This looks reasonable.

There's a bunch of repeated code that I think could be refactored into shared methods, and some comments about using the shared `IMetadata` interface.

Thanks!

macsux commented:
This should be much cleaner now, I've consolidated pretty much all requests into a single method, and added integration tests (currently disabled). 

@brendandburns I've added a test fixture to spin up Kind cluster to do integration testing. Did we wanna add this as part of pipeline? I'm sure we have other tests that can benefit from this.

brendandburns commented on src/KubernetesClient/KubernetesEntityAttribute.cs.orig:
```diff
@@ -0,0 +1,78 @@
+using System;
```
I think this file should be deleted?

brendandburns commented:
This generally looks good to me. It needs a rebase and I think there is an extraneous file that was left over from development.

brendandburns commented:
Oh, and I think the Kind testing is awesome!

macsux commented:
Please hold off on merging this just yet, as if there's a green light to do what is propsed in https://github.com/kubernetes-client/csharp/issues/449 I would like to refactor this as it's own class with extension methods for KubernetesClient. The target implementation would look like this (and removed from Steeltoe if accepted as part of KubernetesClient) if the above is approved:

https://github.com/steeltoeoss-incubator/kubernetes/blob/informers/src/KubernetesBase/Kubernetes.Generic.cs
